### DISCUSSION TITLE
Can someone write a step-by-step guide for compiling llama.cpp with openBLAS on Windows?

### DISCUSSION THREAD

**Discussion by Floating-Pin** [ggml-org/llama.cpp#1153] [Q&A] [answered]
It took me an hour, but I couldn't get it to work. Need help. Also, if openBLAS improves performance, even if not much, why it isn't used by default?

**Comment by SlyEcho** [comment, marked as answer]
I have done this, I'll try to explain.

First, get [w64devkit](https://github.com/skeeto/w64devkit/releases) `w64devkit-1.18.0.zip`

Then, get [OpenBLAS](https://github.com/xianyi/OpenBLAS/releases) `OpenBLAS-0.3.23-x64.zip`

Put w64devkit somewhere you like, no need to set up anything else like PATH, there is just one executable that opens a shell, from there you can build llama.cpp with `make` as usual.

Now take the OpenBLAS release and from there copy `lib/libopenblas.a` into `w64devkit/x86_64-w64-mingw32/lib` and from `include` copy all the .h files to `w64devkit/x86_64-w64-mingw32/include`

`make LLAMA_OPENBLAS=1` should work now.

The files are going to be big because of static linking, you can also copy `libopenblas.dll.a` but then you have to keep the .dll file around where OS can load it.

You can also compile OpenBLAS yourself, for that I recommend the Fortran version of w64devkit.

**Reply by Floating-Pin** [reply]
First, thank you for responding!
But it's weird, I get this error - "ggml.c:149:10: fatal error: cblas.h: No such file or directory
  149 | #include <cblas.h>"
Although I have exactly copied cblas.h and other .h files to w64devkit\include.
Any suggestions?

**Reply by SlyEcho** [reply]
You have to put it to the `x86_64-w64-mingw32/include` directory and the library to `x86_64-w64-mingw32/lib`
It will not work otherwise.

**Reply by Floating-Pin** [reply]
Bah! Dumb me! Ok, now it compiling, but crush right after start :(
main.exe, which I compile with a simple make command, works fine.
Sorry to bother you again.

**Reply by SlyEcho** [reply]
Just in case, do `make clean` before running make with other settings.

**Reply by Floating-Pin** [reply]
No, it didn't help :(
Well, I guess I'm not destined to enjoy openBLAS performance.
Anyway, thanks for your help.

**Reply by SlyEcho** [reply]
Is there some kind of error? Are you running from the same w64devkit shell?

You could try building OpenBLAS yourself, it's not that difficult, just `make` pretty much.

**Reply by Floating-Pin** [reply]
Yes I use w64devkit.exe, then cd to llama.cpp, then make LLAMA_OPENBLAS=1. Several warnings pop up during build, but nothing suspicious. During startup it just crashes (I use powershell to run it).

**Reply by Floating-Pin** [reply]
YES!!! I made it work! I deleted libopenblas.dll.a from x86_64-w64-mingw32/lib, compiled with make LLAMA_OPENBLAS=1, and placed libopenblas.dll from OpenBLAS-0.3.23-x64\bin next to main.exe .

**Reply by SlyEcho** [reply]
If you deleted `libopenblas.dll.a` then the .dll file should not be needed. `main.exe` is about 33 megs in size then.
When it's using the .dll, it is around 2.6 megs + 48 megs for the .dll.

There is a way to control this with the `LDFLAGS` variable to override the makefile:
```sh
# static openblas
make -B LLAMA_OPENBLAS=1 LDFLAGS="C:\apps\w64devkit\x86_64-w64-mingw32\lib\libopenblas.a" main
# libopenblas.dll
make -B LLAMA_OPENBLAS=1 LDFLAGS="C:\apps\w64devkit\x86_64-w64-mingw32\lib\libopenblas.dll.a" main
```

**Reply by Floating-Pin** [reply]
Oh, I see, I didn't copy the .dll library for the exe that needed it, but I copied it for the one that didn't need it. Shame on me. Thank you for your patience.
